### PR TITLE
Trash overgrown schedulers with many usage queues

### PR DESCRIPTION
**Pull request by ryoqun**
#### Problem

From https://github.com/anza-xyz/agave/pull/1211:

> Specifically, there are following several termination conditions which it must handle respectively:
> 
> ...
> 4. Because `UsageQueueLoader` doesn't never evict unused entries. it can become too big with many entries. Its design relies on some external mechanism to mitigate the unbounded growth. Thus, the unified scheduler needs to drop the scheduler itself depending on the size of `UsageQueueLoader`.
> ...

#### Summary of Changes

This pr addresses (4).

### CODE REVIEW

File: unified-scheduler-pool/src/lib.rs [hunk 1]
```diff
@@ -108,6 +109,21 @@ pub type DefaultSchedulerPool =
 
 const DEFAULT_POOL_CLEANER_INTERVAL: Duration = Duration::from_secs(10);
 const DEFAULT_MAX_POOLING_DURATION: Duration = Duration::from_secs(180);
+// Rough estimate of max UsageQueueLoader size in bytes:
+//   UsageFromTask * UsageQeueue's capacity * DEFAULT_MAX_USAGE_QUEUE_COUNT
```
Comment (apfitzge) [inline unified-scheduler-pool/src/lib.rs]: typo:
```suggestion
//   UsageFromTask * UsageQueue's capacity * DEFAULT_MAX_USAGE_QUEUE_COUNT
```

File: unified-scheduler-pool/src/lib.rs [hunk 2]
```diff
@@ -108,6 +109,21 @@ pub type DefaultSchedulerPool =
...
+//   16 bytes      * 128 items              * 262_144 entries               == 512 MiB
+// It's expected that there will be 2 or 3 pooled schedulers constantly when running against
+// mainnnet-beta. That means the total memory consumption for the idle close-to-be-trashed pooled
+// schedulers is set to high: 1.0 ~ 1.5 GiB. This value is chosen to maximize performance under the
```
Comment (apfitzge) [inline unified-scheduler-pool/src/lib.rs]: the "high" reads a bit confusingly to me. maybe justs "is set to 1.0 ~ 1.5 GiB"

File: unified-scheduler-pool/src/lib.rs [hunk 3]
```diff
@@ -648,6 +671,10 @@ where
     }
 
     fn is_trashed(&self) -> bool {
```
Comment (apfitzge) [inline unified-scheduler-pool/src/lib.rs:673]: so want to make a clarification here in the review, doesn't necessarily need a code comment.

This is **only** called after the scheduler is terminated due to abort or successful completion of the slot, is that right?
I want to make sure that the limits from `is_overgrown` will not cause us to skip a slot if such a large one comes through.

**Review by apfitzge — COMMENTED** [review]
minor nit and clarification

File: unified-scheduler-pool/src/lib.rs (hunk 1 above)
Comment (ryoqun) [reply unified-scheduler-pool/src/lib.rs]: oops: https://github.com/anza-xyz/agave/pull/1672/commits/e4bb046aad4e69cd09bcec98b2a5ae2fa2424d9c

File: unified-scheduler-pool/src/lib.rs (hunk 2 above)
Comment (ryoqun) [reply unified-scheduler-pool/src/lib.rs]: thanks: https://github.com/anza-xyz/agave/pull/1672/commits/e4bb046aad4e69cd09bcec98b2a5ae2fa2424d9c

File: unified-scheduler-pool/src/lib.rs (hunk 3 above)
Comment (ryoqun) [reply unified-scheduler-pool/src/lib.rs:673]: > This is **only** called after the scheduler is terminated due to abort or successful completion of the slot, is that right?

yes. it's only called by `::return_to_pool()`.

> I want to make sure that the limits from `is_overgrown` will not cause us to skip a slot if such a large one comes through.

this is true,  again.

**Review by apfitzge — APPROVED** [review]
lgtm

**Pull request merged**